pkg/artifact/job.go

```diff
@@ -0,0 +1,99 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package artifact
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/abcxyz/github-metrics-aggregator/pkg/version"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/workerpool"
+)
+
+// ExecuteJob runs the ingestion pipeline job to read GitHub
+// action workflow logs from GitHub and store them into GCS.
+func ExecuteJob(ctx context.Context, cfg *Config) error {
+	logger := logging.FromContext(ctx)
+
+	bq, err := NewBigQuery(ctx, cfg.ProjectID, cfg.DatasetID)
+	if err != nil {
+		return fmt.Errorf("failed to create bigquery client: %w", err)
+	}
+	defer bq.Close()
+
+	eventsTableDotNotation := formatGoogleSQL(cfg.ProjectID, cfg.DatasetID, cfg.EventsTableID)
+	artifactsTableDotNotation := formatGoogleSQL(cfg.ProjectID, cfg.DatasetID, cfg.ArtifactsTableID)
+
+	// Create a pool of workers to manage all of the log ingestions
+	pool := workerpool.New[ArtifactRecord](&workerpool.Config{
+		Concurrency: int64(runtime.NumCPU()),
+		StopOnError: false,
+	})
+
+	// Setup a log ingester to process ingestion events
+	logsFn, err := NewLogIngester(ctx, cfg.BucketName, cfg.GitHubAppID, cfg.GitHubInstallID, cfg.GitHubPrivateKey)
+	if err != nil {
+		return fmt.Errorf("failed to create log ingester: %w", err)
+	}
+
+	logger.InfoContext(ctx, "ingestion job starting",
+		"name", version.Name,
+		"commit", version.Commit,
+		"version", version.Version)
+
+	// Read up to `BatchSize` number of events that need to be processed
+	query := fmt.Sprintf(SourceQuery, eventsTableDotNotation, artifactsTableDotNotation, cfg.BatchSize)
+	events, err := Query[EventRecord](ctx, bq, query)
+	if err != nil {
+		return fmt.Errorf("failed to query bigquery for events: %w", err)
+	}
+
+	// Fan out the work of processing all of the events that were found
+	for _, event := range events {
+		if err := pool.Do(ctx, func() (ArtifactRecord, error) {
+			artifact := logsFn.ProcessElement(ctx, *event)
+			// Errors are handled by the element processor and are flagged as special
+			// artifact records. There is no possible error returned from processing.
+			return artifact, nil
+		}); err != nil {
+			return fmt.Errorf("failed to submit job to worker pool: %w", err)
+		}
+	}
+	// When all of the workers are complete, extract the result values
+	results, err := pool.Done(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to ingest logs for events: %w", err)
+	}
+	artifacts := Map(func(r *workerpool.Result[ArtifactRecord]) ArtifactRecord { return r.Value }, results)
```

Comment:
This is a bit weird and not idiomatic to Go. I would expect:

```go
artifacts := make([]*ArtifactRecord, 0, len(results))
for _, v := range results {
  artifacts = append(artifacts, v.Value)
}
```

and then remove the `Map` helper.

Reply:
Why ~do you go people~ does Go have such a dislike for the Map pattern that's available in every other language? I'm genuinely curious. I don't mind making this change but it seems like such a strange omission that I really want to know the reasoning.

Reply:
It's nearly 4x slower, even if I force the function to inline.

```text
goos: darwin
goarch: arm64
pkg: github.com/sethvargo/bench
BenchmarkDirect-10    	1000000000	         1.656 ns/op	       8 B/op	       0 allocs/op
BenchmarkMap-10       	1000000000	         4.558 ns/op	       8 B/op	       0 allocs/op
```

Reply:
That's unfortunate - is it the function call that makes it slower, the generics? Its basically the same code.

Not a problem to solve today but I'm invested in the `why` of this. 

I'll make the change you requested.